.maintain/gitlab/check_companion_build.sh

```diff
@@ -0,0 +1,123 @@
+# (since it doesn't have the changes to polkadot that were required in the polkadot
+# companion PR)
+
+# So invoke this script like (arguments in [] indicate optional arguments)
+# ./check_companion_build.sh paritytech cumulus ['cargo test --release'] [paritytech/polkadot]
+
+set -e
+
+# Include the common functions library
+#shellcheck source=../common/lib.sh
+. "$(dirname "${0}")/../common/lib.sh"
+
+ORGANISATION=$1
+REPO=$2
+BUILDSTRING=${3:-cargo test --workspace --release}
+DEPS=("${@:4}")
+SUBSTRATE_DIR="$(pwd)"
+
+boldprint () { printf "|\n| \033[1m%s\033[0m\n|\n" "${@}"; }
+boldcat () { printf "|\n"; while read -r l; do printf "| \033[1m%s\033[0m\n" "${l}"; done; printf "|\n" ; }
+
+boldcat <<-EOT
+
+
+check_${REPO}_companion_build
+==============================
+
+this job checks if there is a string in the description of the pr like
+
+$REPO companion: $ORGANISATION/$REPO#567
+
+
+it will then run cargo check from this ${REPO}'s branch with substrate code
+from this pull request. otherwise, it will uses master instead
+
+
+EOT
+
+# Set the user name and email to make merging work
+git config --global user.name 'CI system'
+git config --global user.email '<>'
+
+# Merge master into our branch before building to make sure we don't miss
+# any commits that are required.
+git fetch --depth 100 origin
+git merge origin/master
+
+# Clone the current master branch into a local directory
+# NOTE: we need to pull enough commits to be able to find a common
+# ancestor for successfully performing merges below.
+git clone --depth 20 "https://github.com/${ORGANISATION}/${REPO}.git"
+
+cd "$REPO"
+
+# either it's a pull request then check for a companion otherwise use
+# master
+# shellcheck disable=SC2003
+if expr match "${CI_COMMIT_REF_NAME}" '^[0-9]\+$' >/dev/null
+then
+  boldprint "this is pull request no ${CI_COMMIT_REF_NAME}"
+  pr_companion="$(get_companion "paritytech/substrate" "$CI_COMMIT_REF_NAME" "$ORGANISATION/$REPO")"
+  if [ "$pr_companion" ]
+  then
+    boldprint "companion pr specified/detected: #${pr_companion}"
+    git fetch origin "refs/pull/${pr_companion}/head:pr/${pr_companion}"
+    git checkout "pr/${pr_companion}"
+    git merge origin/master
+  else
+    boldprint "no companion branch found - building ${REPO}:master"
+  fi
+
+  # If this repo has any additional dependencies, we should check whether they
+  # are mentioned as companions as well. If they are, we patch this repo to
+  # use that companion build as well. See example at top of this script
+  # Note: Will only work with repos supported by diener
+  declare -A diener_commands
+  diener_commands=()
+  diener_commands["paritytech/polkadot"]='--polkadot'
+  diener_commands["paritytech/substrate"]='--substrate'
+  diener_commands["paritytech/grandpa-bridge-gadget"]='--beefy'
+
+  for dep in "${DEPS[@]}"; do
+    dep_companion="$(get_companion "paritytech/substrate" "$CI_COMMIT_REF_NAME" "$dep")"
```

Comment:
since you're already fetching the description once in the lines above, it would be good to reuse it instead of making more API requests